### DISCUSSION TITLE
Performance Benchmarking

### DISCUSSION THREAD

**Discussion by AlexandruGG** [loophp/collection#160] [Ideas]
@drupol opening this so I don't forget. I think we should introduce performance benchmarking with https://github.com/phpbench/phpbench.

We can detect which operations are more inefficient than we'd like and target them for refactoring. Also this would ensure that future big changes (like the introduction of FPT toolbox) don't negatively affect performance too much.

One example I noticed [Duplicate](https://github.com/loophp/collection/blob/master/src/Operation/Duplicate.php) could be improved by using a "hash map"-style array instead of an ordered list (though there would have to be some processing because of php only allowing `int|string` for keys).

What do you think?

**Comment by drupol** [comment]
I definitely support this. I already have a couple of tests using that framework on my local instances and they were never committed.

I'm wondering if there is a way to publish the results of the tests online and compare the numbers when doing PRs.

**Reply by AlexandruGG** [reply]
I've never used that tool before but based on the docs looks like we have both [Assertions](https://phpbench.readthedocs.io/en/latest/guides/assertions.html) and [Regression Testing](https://phpbench.readthedocs.io/en/latest/guides/regression-testing.html) which we can set up to do what we want

**Reply by drupol** [reply]
Ah this is something I never tried, and it looks amazing :)

**Comment by drupol** [comment]
> One example I noticed [Duplicate](https://github.com/loophp/collection/blob/master/src/Operation/Duplicate.php) could be improved by using a "hash map"-style array instead of an ordered list (though there would have to be some processing because of php only allowing `int|string` for keys).

Mmmh if you have a better option for `Duplicate`, I'm curious to see what you have in mind.

**Reply by AlexandruGG** [reply]
https://gist.github.com/AlexandruGG/53cad181f5242179148cdba5e7ed1576 I'm thinking something on the lines of this.

Of course it would have to be more complex in order to allow using values that are not `int|string` as keys, but we could use [spl_object_hash](https://www.php.net/manual/en/function.spl-object-hash.php) for objects for example and use that as the key.

**Reply by drupol** [reply]
I think there's nothing faster in PHP than a array indexed by integers, I think doing what your propose will be slower. (_I think I saw it from this presentation: https://www.youtube.com/watch?v=3mcGzBnSm1c_)

Anyway, we can do some bench tests and see how it goes for sure.

**Reply by AlexandruGG** [reply]
> I think there's nothing faster in PHP than a array indexed by integers, I think doing what your propose will be slower. (_I think I saw it from this presentation: https://www.youtube.com/watch?v=3mcGzBnSm1c_)
> 
> Anyway, we can do some bench tests and see how it goes for sure.

I think I saw that presentation but don't remember that bit.

In any case I had already tested the gist and it's way, way, way faster. Just try it with a Collection of 100,000 elements. `in_array` has a O(N) runtime whereas the hash map approach will be closer to constant time O(1). Of course I haven't tested it yet with all the logic that would be needed to deal with elements that are not int|string and so cannot be used as keys

**Reply by drupol** [reply]
Of course it will be faster if we pretend that keys are only `int|string` but it might not always be the case.

Adding conditions and `spl_object_hash` and stuff like these are not what I want to introduce in the library, we must keep it simple and clean, and we know that there is a price to pay in the performances.

**Reply by AlexandruGG** [reply]
> Of course it will be faster if we pretend that keys are only `int|string` but it might not always be the case.
> 
> Adding conditions and `spl_object_hash` and stuff like these are not what I want to introduce in the library, we must keep it simple and clean, and we know that there is a price to pay in the performances.

I know what you mean. In any case for now it's fine, until someone has a real-world issue with `Duplicate` we don't need to do anything with it specifically :)
